server/cli/src/test_connection.rs

```diff
@@ -0,0 +1,499 @@
+use reqwest::{Client, Url};
+use serde::{Deserialize, Serialize};
+use server::configuration;
+use service::{
+    apis::login_v4::{LoginApiV4, LoginInputV4, LoginUserTypeV4},
+    app_data::{AppDataService, AppDataServiceTrait},
+    service_provider::ServiceProvider,
+    settings::is_develop,
+    sync::{
+        api::{SyncApiSettings, SyncApiV5},
+        api_v6::SyncApiV6,
+        settings::SyncSettings,
+    },
+};
+use tokio::sync::mpsc;
+
+#[derive(Parser, Debug)]
+struct Args {
+    #[arg(short, long)]
+    gui: bool,
+    #[arg(short, long)]
+    username: Option<String>,
+    #[arg(short, long)]
+    password: Option<String>,
+    #[arg(short, long)]
+    server_uuid: Option<String>,
+}
+
+#[tokio::main]
+async fn main() {
+    simple_logger::init_with_level(Level::Info).unwrap();
+
+    let args = Args::parse();
+    let gui = args.gui;
+
+    let (gui_tx, gui_rc) = mpsc::channel(10);
+    let test_task = tokio::spawn(perform_tests(gui_tx, args));
+
+    if gui {
+        let options = eframe::NativeOptions {
+            viewport: egui::ViewportBuilder::default().with_inner_size([320.0, 240.0]),
+            ..Default::default()
+        };
+
+        eframe::run_native(
+            "OMS - Test Connections",
+            options,
+            Box::new(|cc| {
+                egui_extras::install_image_loaders(&cc.egui_ctx); // This gives us image support
+                Box::new(Gui::new(gui_rc))
+            }),
+        )
+        .unwrap();
+    }
+
+    test_task.await.unwrap();
+}
+
+async fn perform_tests(gui_tx: mpsc::Sender<GuiState>, args: Args) {
+    let gui = args.gui;
+    let mut test_data = TestData {
+        server_config: None,
+        sync_api_v5: None,
+        args,
+    };
+
+    let tests: Vec<Box<dyn Test + Send>> = vec![
+        Box::new(ConfigTest),
+        Box::new(PingTest),
+        Box::new(DatabaseTest),
+        Box::new(LoginTest),
+        Box::new(SyncTest),
+        Box::new(SyncV6Test),
+    ];
+
+    let mut gui_state = GuiState {
+        tests: vec![
+            ("Config".to_string(), TestState::Pending),
+            ("Ping".to_string(), TestState::Pending),
+            ("Database".to_string(), TestState::Pending),
+            ("Login".to_string(), TestState::Pending),
+            ("Sync V5".to_string(), TestState::Pending),
+            ("Sync V6".to_string(), TestState::Pending),
+        ],
+    };
+
+    for i in 0..tests.len() {
+        if gui {
+            gui_state.tests[i].1 = TestState::Running;
+            gui_tx.send(gui_state.clone()).await.unwrap();
+        }
+
+        let result = tests[i].run(&mut test_data).await;
+        match &result {
+            Ok(msg) => {
+                info!("{} test passed: {}", gui_state.tests[i].0, msg);
+            }
+            Err(msg) => {
+                error!("{} test failed: {}", gui_state.tests[i].0, msg);
+            }
+        }
+
+        if gui {
+            gui_state.tests[i].1 = match result {
```

Comment:
not a fan of indexing an array by another array, i think we need to use type safety of rust as much as possible (this can really be one array I think, with test implementing a method returning string in tuple)